stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -104,40 +110,74 @@ struct reduce_sum_impl<ReduceFunction, require_var_t<ReturnType>, ReturnType,
         local_sub_slice.emplace_back(deep_copy_vars(vmapped_[i]));
       }
 
-      // Create nested autodiff copies of all shared arguments that do not point
-      //   back to main autodiff stack
-      auto args_tuple_local_copy = apply(
-          [&](auto&&... args) {
-            return std::tuple<decltype(deep_copy_vars(args))...>(
-                deep_copy_vars(args)...);
-          },
-          args_tuple_);
-
-      // Perform calculation
-      var sub_sum_v = apply(
-          [&](auto&&... args) {
-            return ReduceFunction()(local_sub_slice, r.begin(), r.end() - 1,
-                                    msgs_, args...);
-          },
-          args_tuple_local_copy);
-
-      // Compute Jacobian
-      sub_sum_v.grad();
-
-      // Accumulate value of reduce_sum
-      sum_ += sub_sum_v.val();
-
-      // Accumulate adjoints of sliced_arguments
-      accumulate_adjoints(sliced_partials_ + r.begin() * num_vars_per_term_,
-                          std::move(local_sub_slice));
-
-      // Accumulate adjoints of shared_arguments
-      apply(
-          [&](auto&&... args) {
-            accumulate_adjoints(args_adjoints_.data(),
-                                std::forward<decltype(args)>(args)...);
-          },
-          std::move(args_tuple_local_copy));
+      if (std::this_thread::get_id() == main_thread_) {
+        // on the main thread we can use the input operands given as
+        // input to reduce_sum (but need to set their adjoints to
+        // zero)
+
+        // Perform calculation
+        var sub_sum_v = apply(
+            [&](auto&&... args) {
+              return ReduceFunction()(local_sub_slice, r.begin(), r.end() - 1,
+                                      msgs_, args...);
+            },
+            args_tuple_);
+
+        // Compute Jacobian
+        sub_sum_v.grad();
+
+        // Accumulate value of reduce_sum
+        sum_ += sub_sum_v.val();
+
+        // Accumulate adjoints of sliced_arguments
+        accumulate_adjoints(sliced_partials_ + r.begin() * num_vars_per_term_,
+                            std::move(local_sub_slice));
+
+        // Accumulate adjoints of shared_arguments and set adjoints to zero
+        apply(
+            [&](auto&&... args) {
+              accumulate_adjoints(args_adjoints_.data(), args...);
+              zero_adjoints(args...);
+            },
+            args_tuple_);
+
+      } else {
+        // Create nested autodiff copies of all shared arguments that do not
+        // point
+        //   back to main autodiff stack
+        auto args_tuple_local_copy = apply(
+            [&](auto&&... args) {
+              return std::tuple<decltype(deep_copy_vars(args))...>(
+                  deep_copy_vars(args)...);
```

Comment:
Good point with this whole pull, but now I'm wondering why do we ever do these copies? Isn't it always safe to accumulate back into the main thread varis because there is no race condition?

Reply:
At least the varis here in the sliced argument.

Even if someone passed the same variable as a sliced argument and shared argument, it is always copied for the shared argument anyway, so still no race condition.

Reply:
Yes, absolutely correct. We can avoid the deep copy for the sliced argument as well. That will improve performance quite a bit more.

It's just that this change makes the assumption that the adjoints of the passed in var's are 0. Moreover, we modify things on the main AD tape, since the vari's are being used. I think this is all OK to do, but maybe needs a mention in the doc.

I will try to benchmark the latest change which recycles the sliced arguments vari's. That should make quite a difference is my guess.

It's cool to see that there is more to get from `reduce_sum`! With the scoping stuff this thing just got a decent boost.

Reply:
> It's just that this change makes the assumption that the adjoints of the passed in var's are 0

Oh yeah, we should probably save the adjoints, set them to zero, and then restore them after we're done then. That's a bit tricky and I'm not sure if I've forgotten anything.

Reply:
What about we check the vari's to be zero and throw an exception if they are not 0?

I am running benchmarks right now and using this trick for the slicing argument gives us a 10% speedup when I see that right for a bernoulli example which suffers a lot from memory stuff.